### ISSUE TITLE
Using pre-trained model weights and re-training further causes degradation

### ISSUE THREAD

**Issue by GeorgiAngelov** [rbgirshick/py-faster-rcnn#352]
I am working on further training a model and using it as the weights for training. The model was pretty good at recognizing the objects I wanted it to recognize but after I did an additional 30k iterations and by renaming the last 2 InnerProduct layers, I now get really low confidence detection and very random ones.

Is there a guide on the proper way to train using pre-trained model for the weights ? My data only contains 5 of the 21 classes so that is why I renamed the last 2 InnerProduct layers so it can get re-initialized for the new data.

Should I change the weight_decay ?

**Comment by hgaiser** [comment]
https://www.google.nl/search?q=1%2F21&oq=1%2F21&aqs=chrome..69i57j69i64.561j0j7&sourceid=chrome&ie=UTF-8

I doubt that is a coincidence :)

I am guessing your test.prototxt or train.prototxt isn't updated properly to use 5 classes. Could you share them so I can have a look at it?

**Comment by GeorgiAngelov** [comment]
@hgaiser, that's what I figured.... I just updated my post to include the test and train prototxt files. Thanks for responding!

**Comment by hgaiser** [comment]
They appear correct at a glance. Did you also modify the following:

https://github.com/rbgirshick/py-faster-rcnn/blob/master/lib/datasets/pascal_voc.py#L30

It should also contain 6 classes, preferably with the correct names.

**Comment by GeorgiAngelov** [comment]
@hgaiser , i did change the pascal_voc accordingly.

**Comment by hgaiser** [comment]
Then I am a bit out of ideas, sorry.

**Comment by hgaiser** [comment]
A bit of a wild guess, but did you try deleting the cache?

**Comment by GeorgiAngelov** [comment]
@hgaiser, yeah I did... What I ended up doing is I reverted the changes I made to the .prototxt files regarding the number of classes and such. Also, I believe there was a piece of code somewhere that still expected all 21 classes but it wasn't in a way that caused a crash. I am wondering if that was the case..

**Comment by brisker** [comment]
@GeorgiAngelov
Have you solved this problem?

**Comment by GeorgiAngelov** [comment]
@brisker, I have still yet to find why the confidence is constant.... It only happens when I change my class count... I am still looking for an answer to this... For the time being, I simply kept all 21 classes of the VGG16 network.

**Comment by brisker** [comment]
is it ok when we have  2 classes but still write 21 in the prototxt?


On 10/05/2016 12:02, Georgi Angelov wrote:

@brisker, I have still yet to find why the confidence is constant.... It only happens when I change my class count... I am still looking for an answer to this... For the time being, I simply kept all 21 classes of the VGG16 network.

**Comment by GeorgiAngelov** [comment]
As far as I know, if you leave everything the same, you can still further train the network by using the weights and only further train it on 2 classes. I have a question posted in the Caffe repo regarding this to see if it is really detrimental to modify the network: https://github.com/BVLC/caffe/issues/4787
